File: scripts/scrape_videos.py
#!/usr/bin/env python3
"""
Viralkand.com video scraper - extracts direct MP4 links
"""

import json
import requests
from bs4 import BeautifulSoup
from datetime import datetime, timezone
import re
import time
import os

# Target listing pages to scrape
LISTING_PAGES = [
    "https://viralkand.com",
]

def get_video_post_urls(listing_url):
    """Get all post URLs from a listing page."""
    try:
        headers = {
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
        }
        
        print(f"\n📄 Fetching: {listing_url}")
        resp = requests.get(listing_url, headers=headers, timeout=30)
        resp.raise_for_status()
        
        soup = BeautifulSoup(resp.text, "html.parser")
        post_urls = []
        
        # Look for post links (skip category/tag/page links)
        for link in soup.find_all("a", href=True):
            href = link["href"]
            # Only viralkand.com posts
            if "viralkand.com" in href and all(skip not in href for skip in ["/page/", "/category/", "/tag/", "/author/"]):
                if href.endswith("/") and href not in post_urls:
                    post_urls.append(href)
        
        print(f"   Found {len(post_urls)} post links")
        return post_urls
        
    except Exception as e:
        print(f"   Error: {e}")
        return []


def extract_video_from_post(post_url):
    """Extract video data from a post page."""
    try:
        headers = {
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
            "Referer": "https://viralkand.com",
        }
        
        print(f"  🔍 {post_url}")
        resp = requests.get(post_url, headers=headers, timeout=30)
        resp.raise_for_status()
        
        soup = BeautifulSoup(resp.text, "html.parser")
        
        # Extract title
        title = "Untitled Video"
        og_title = soup.find("meta", property="og:title")
        if og_title and og_title.get("content"):
            title = og_title["content"][:100]
        
        # Extract thumbnail
        thumbnail = ""
        og_image = soup.find("meta", property="og:image")
        if og_image and og_image.get("content"):
            thumbnail = og_image["content"]
        
        # Find video embed
        embed_url = None
        
        # Strategy 1: Look for iframe
        for iframe in soup.find_all("iframe"):
            src = iframe.get("src")
            if not src:
                continue
            # Skip ads
            if any(ad in src.lower() for ad in ["doubleclick", "googlesyndication", "adserver", "ads"]):
                continue
            embed_url = src if src.startswith("http") else ("https:" + src if src.startswith("//") else "")
            break
        
        # Strategy 2: Look for HTML5 video tag (direct MP4)
        if not embed_url:
            video_tag = soup.find("video")
            if video_tag:
                source = video_tag.find("source")
                if source and source.get("src"):
                    embed_url = source["src"]
        
        if not embed_url:
            print(f"     ❌ No video found")
            return None
        
        # Generate ID from URL slug
        slug_match = re.search(r"/([^/]+)/$", post_url)
        video_id = slug_match.group(1) if slug_match else f"vid-{abs(hash(post_url)) % 100000}"
        
        print(f"     ✅ {title[:50]}")
        
        return {
            "id": f"vid-{video_id[:50]}",
            "title": title,
            "description": "Viral video from viralkand.com",
            "category": "Viral",
            "duration": "00:00",
            "embedUrl": embed_url,
            "thumbnailUrl": thumbnail,
            "tags": ["viralkand", "viral"],
            "uploadedAt": datetime.now(timezone.utc).isoformat(),
            "views": 0
        }
        
    except Exception as e:
        print(f"     ❌ Error: {e}")
        return None


def scrape_all():
    """Main scraper function."""
    print("=" * 70)
    print("🚀 VIRALKAND SCRAPER STARTING")
    print("=" * 70)
    
    os.makedirs("data", exist_ok=True)
    
    all_videos = []
    
    # Step 1: Collect all post URLs
    all_post_urls = []
    for listing_url in LISTING_PAGES:
        urls = get_video_post_urls(listing_url)
        all_post_urls.extend(urls)
        time.sleep(2)  # Be polite
    
    all_post_urls = list(set(all_post_urls))  # Remove duplicates
    print(f"\n📊 Total posts found: {len(all_post_urls)}")
    
    # Step 2: Extract videos from each post (limit 30 per run)
    for i, post_url in enumerate(all_post_urls[:30], 1):
        print(f"\n[{i}/30]")
        video = extract_video_from_post(post_url)
        if video:
            all_videos.append(video)
        time.sleep(1.5)  # Be polite
    
    # Remove duplicate embeds
    seen = set()
    unique_videos = []
    for v in all_videos:
        if v["embedUrl"] not in seen:
            seen.add(v["embedUrl"])
            unique_videos.append(v)
    
    print("\n" + "=" * 70)
    print(f"✅ SUCCESS!")
    print(f"   Scraped: {len(unique_videos)} unique videos")
    print("=" * 70)
    
    # Save to JSON
    with open("data/videos.json", "w", encoding="utf-8") as f:
        json.dump(unique_videos, f, ensure_ascii=False, indent=2)
    
    print("\n💾 Saved to data/videos.json")


if __name__ == "__main__":
    scrape_all()

File: scripts/scrape-videos.js
import fs from "fs";
import path from "path";
import { chromium } from "playwright";

/* ===================== CONFIG ===================== */

const BASE = "https://desimyhub.net";
const PAGES = [
  `${BASE}/latest/`,
  `${BASE}/latest/page/2/`,
];

/* Blocked ad / junk domains */
const BLOCKED_SRC = [
  "ads",
  "doubleclick",
  "googlesyndication",
  "afcdn",
  "pop",
  "traffic",
];

/* ===================== FILE ===================== */

const DATA_PATH = path.resolve("data/videos.json");

/* ===================== HELPERS ===================== */

function loadExisting() {
  try {
    return JSON.parse(fs.readFileSync(DATA_PATH, "utf8"));
  } catch {
    return [];
  }
}

function save(videos) {
  fs.writeFileSync(DATA_PATH, JSON.stringify(videos, null, 2));
}

function makeId(str) {
  return "vid-" + Buffer.from(str).toString("base64").slice(0, 32);
}

function isBlocked(url = "") {
  return BLOCKED_SRC.some(b => url.toLowerCase().includes(b));
}

/* ===================== MAIN ===================== */

async function run() {
  console.log("▶ Scraper started");

  const existing = loadExisting();
  const seen = new Set(existing.map(v => v.embedUrl));
  const results = [...existing];

  const browser = await chromium.launch({ headless: true });
  const context = await browser.newContext({
    userAgent:
      "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/120 Safari/537.36",
  });

  const page = await context.newPage();

  /* 🚫 BLOCK ADS / TRACKERS */
  await page.route("**/*", route => {
    const url = route.request().url();
    if (BLOCKED_SRC.some(b => url.includes(b))) {
      return route.abort();
    }
    route.continue();
  });

  for (const url of PAGES) {
    console.log("📄 Listing:", url);

    await page.goto(url, {
      waitUntil: "domcontentloaded",
      timeout: 30000,
    });

    await page.waitForTimeout(2000);

    /* UNIVERSAL CARD SELECTOR */
    let cards = await page.$$("a.video");
    if (!cards.length) cards = await page.$$("article a[href]");
    if (!cards.length) cards = await page.$$(".post a[href]");
    if (!cards.length) cards = await page.$$("a[href]");

    console.log(`Found ${cards.length} cards`);

    for (const card of cards) {
      try {
        const href = await card.getAttribute("href");
        if (!href || href.startsWith("#")) continue;

        const postUrl = href.startsWith("http") ? href : BASE + href;

        /* TITLE */
        const title =
          (await card.$eval("h1", el => el.textContent.trim()).catch(() => null)) ||
          (await card.$eval("h2", el => el.textContent.trim()).catch(() => null)) ||
          (await card.$eval("h3", el => el.textContent.trim()).catch(() => null)) ||
          (await card.getAttribute("title")) ||
          "Video";

        /* THUMB */
        const thumbnail =
          (await card.$eval("img", el => el.src).catch(() => "")) || "";

        /* DURATION */
        const duration =
          (await card
            .$eval(".time, .clock, .duration", el => el.textContent.trim())
            .catch(() => "00:00")) || "00:00";

        /* OPEN POST PAGE */
        const post = await context.newPage();

        await post.route("**/*", route => {
          const url = route.request().url();
          if (BLOCKED_SRC.some(b => url.includes(b))) {
            return route.abort();
          }
          route.continue();
        });

        await post.goto(postUrl, {
          waitUntil: "domcontentloaded",
          timeout: 30000,
        });

        await post.waitForTimeout(2000);

        /* EMBED EXTRACTION */
        let embedUrl = null;

        const iframeUrls = await post.$$eval("iframe", els =>
          els.map(el => el.src).filter(Boolean)
        );

        embedUrl = iframeUrls.find(src => !isBlocked(src)) || null;

        if (!embedUrl) {
          const videoUrls = await post.$$eval("video source", els =>
            els.map(el => el.src).filter(Boolean)
          );
          embedUrl = videoUrls.find(src => !isBlocked(src)) || null;
        }

        await post.close();

        if (!embedUrl || seen.has(embedUrl)) continue;

        results.push({
          id: makeId(embedUrl),
          title,
          description: `Video from ${new URL(BASE).hostname}`,
          category: "Viral",
          duration,
          embedUrl,
          thumbnailUrl: thumbnail,
          tags: [new URL(BASE).hostname],
          uploadedAt: new Date().toISOString(),
          views: 0,
        });

        seen.add(embedUrl);
        console.log("➕ Added:", title.slice(0, 60));
      } catch (err) {
        console.log("⚠️ Skipped one card");
        continue;
      }
    }
  }

  await browser.close();
  save(results);

  console.log(`✅ Done. Total videos: ${results.length}`);
}

/* ⏱ HARD SAFETY EXIT */
setTimeout(() => {
  console.error("⏱ Forced exit to prevent hang");
  process.exit(0);
}, 4 * 60 * 1000);

run();
